presto-main/src/main/java/com/facebook/presto/operator/JoinBridgeManager.java

```diff
@@ -0,0 +1,538 @@
+    {
+        return new JoinBridgeManager<>(
+                false,
+                UNGROUPED_EXECUTION,
+                UNGROUPED_EXECUTION,
+                ignored -> factory,
+                factory.getOutputTypes());
+    }
+
+    private final List<Type> buildOutputTypes;
+    private final boolean buildOuter;
+    private final PipelineExecutionStrategy probeExecutionStrategy;
+    private final PipelineExecutionStrategy buildExecutionStrategy;
+    private final Function<Lifespan, T> joinBridgeProvider;
+
+    private final FreezeOnReadCounter probeFactoryCount = new FreezeOnReadCounter();
+
+    private final AtomicBoolean initialized = new AtomicBoolean();
+    private InternalJoinBridgeDataManager<T> internalJoinBridgeDataManager;
+
+    public JoinBridgeManager(
+            boolean buildOuter,
+            PipelineExecutionStrategy probeExecutionStrategy,
+            PipelineExecutionStrategy lookupSourceExecutionStrategy,
+            Function<Lifespan, T> lookupSourceFactoryProvider,
+            List<Type> buildOutputTypes)
+    {
+        this.buildOuter = buildOuter;
+        this.probeExecutionStrategy = requireNonNull(probeExecutionStrategy, "probeExecutionStrategy is null");
+        this.buildExecutionStrategy = requireNonNull(lookupSourceExecutionStrategy, "lookupSourceExecutionStrategy is null");
+        this.joinBridgeProvider = requireNonNull(lookupSourceFactoryProvider, "joinBridgeProvider is null");
+        this.buildOutputTypes = requireNonNull(buildOutputTypes, "buildOutputTypes is null");
+    }
+
+    private void initializeIfNecessary()
+    {
+        if (!initialized.get()) {
+            synchronized (this) {
+                if (initialized.get()) {
+                    return;
+                }
+                int finalProbeFactoryCount = probeFactoryCount.get();
+                internalJoinBridgeDataManager = internalJoinBridgeDataManager(probeExecutionStrategy, buildExecutionStrategy, joinBridgeProvider, finalProbeFactoryCount, buildOuter ? 1 : 0);
+                initialized.set(true);
+            }
+        }
+    }
+
+    public List<Type> getBuildOutputTypes()
+    {
+        return buildOutputTypes;
+    }
+
+    public PipelineExecutionStrategy getBuildExecutionStrategy()
+    {
+        return buildExecutionStrategy;
+    }
+
+    public void incrementProbeFactoryCount()
+    {
+        probeFactoryCount.increment();
+    }
+
+    public T getJoinBridge(Lifespan lifespan)
+    {
+        initializeIfNecessary();
+        return internalJoinBridgeDataManager.getJoinBridge(lifespan);
+    }
+
+    /**
+     * Invoked when a probe operator factory indicates that it will not
+     * create any more operators, for any lifespan.
+     * <p>
+     * It is expected that this method will only be invoked after
+     * {@link #probeOperatorFactoryClosed(Lifespan)} has been invoked
+     * for every known lifespan.
+     */
+    public void probeOperatorFactoryClosedForAllLifespans()
```

Comment:
Note in the `Operator` interface, it's just two overloaded version of `noMoreOperators`:

```
    /**
     * Declare that createOperator will not be called any more and release
     * any resources associated with this factory.
     * <p>
     * This method will be called only once.
     * Implementation doesn't need to worry about duplicate invocations.
     * <p>
     * It is guaranteed that this will only be invoked after {@link #noMoreOperators(Lifespan)}
     * has been invoked for all applicable driver groups.
     */
    void noMoreOperators();

    /**
     * Declare that createOperator will not be called any more for the specified Lifespan,
     * and release any resources associated with this factory.
     * <p>
     * This method will be called only once for each Lifespan.
     * Implementation doesn't need to worry about duplicate invocations.
     * <p>
     * It is guaranteed that this method will be invoked for all applicable driver groups
     * before {@link #noMoreOperators()} is invoked.
     */
    void noMoreOperators(Lifespan lifespan);
```

What about just use `probeOperatorFactoryClosed` here as well? 

The other option would just name it `noMoreOperatorsForProbeOperatorFactory`